### PR TITLE
Add crop op

### LINKED ISSUE #3905: Add crop op

### PR DESCRIPTION
**Pull request by wanghaoshuang**
fixed #3905

### CODE REVIEW

File: paddle/operators/crop_op.cc [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+};
+
+class CropOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  CropOpMaker(framework::OpProto *proto, framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput("X", "The input of crop op");
+    AddInput("Y", "The input used as reference for cropping. ");
+    AddOutput("Out", "The output of crop op.");
+    AddComment(R"DOC(
```
Comment (dzhwinter) [inline paddle/operators/crop_op.cc]: This is a bad documentation, `The offsets for cropping` contains no information at all. Maybe describe some detail like (not necessary) 
1. in which scene this Op will be used.
2. Input, Output of the Op.
3. the detail of this attribute, such as `shape` refers to the area shape of interest in the source image.

File: paddle/operators/crop_op.h [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+#include "paddle/framework/op_registry.h"
+
+namespace paddle {
+namespace operators {
+
+template <typename T, size_t D, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenTensor = framework::EigenTensor<T, D, MajorType, IndexType>;
+
+using Tensor = framework::Tensor;
```
Comment (dzhwinter) [inline paddle/operators/crop_op.h]: follow the name convention of the google style [click here in detail](https://google.github.io/styleguide/cppguide.html#Namespaces). we should not use any `using` in the header file.

File: paddle/operators/crop_op.cc (hunk 1 above)
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cc]: Thx. I will refine it.

File: paddle/operators/crop_op.h (hunk 2 above)
Comment (wanghaoshuang) [reply paddle/operators/crop_op.h]: Thx.

Comment (dzhwinter) [reply paddle/operators/crop_op.h]: In addition, Tensor has no template parameter, Just write
```c++
using framework::Tensor;
```
is okay.

File: paddle/operators/crop_op.h [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+template <typename Place, typename T, size_t D>
+void CropFunction(const framework::ExecutionContext& context) {
+  auto* x = context.Input<Tensor>("X");
+  auto* out = context.Output<Tensor>("Out");
+  out->mutable_data<T>(context.GetPlace());
+  auto x_dims = x->dims();
+  auto out_dims = out->dims();
+
+  auto offsets = context.op().GetAttr<std::vector<int>>("offsets");
+  PADDLE_ENFORCE_EQ(
```
Comment (dzhwinter) [inline paddle/operators/crop_op.h]: Please move this enforce check into `Infershape`. 
Operator's `Infershape` will be run before runtime, that's where real config check happens.

File: paddle/operators/crop_op.cc [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+namespace operators {
+
+using framework::Tensor;
+
+class CropOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
```
Comment (qingqing01) [inline paddle/operators/crop_op.cc]: `InferShape`里的检查需要全面一些，比如对输入非空的检查： https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/squared_l2_distance_op.cc#L26

File: paddle/operators/crop_op.cc [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    if (Y == nullptr) {
+      auto shape = GetAttr<std::vector<int>>("shape");
+      PADDLE_ENFORCE_EQ(
+          shape.size(), dim0.size(),
+          "Shape size should be equal to dimention size of input tensor.");
+      std::vector<int64_t> tensor_shape(shape.size());
+      for (int i = 0; i < shape.size(); ++i) {
+        tensor_shape[i] = (int64_t)shape[i];
+      }
+      ctx.Output<Tensor>("Out")->Resize(
```
Comment (qingqing01) [inline paddle/operators/crop_op.cc]: LoDTensor合入后，InferShape的Output里要使用：`Output<framework::LoDTensor>`

File: paddle/operators/crop_op.cc [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+          paddle::framework::make_ddim(tensor_shape));
+    } else {
+      ctx.Output<Tensor>("Out")->Resize(Y->dims());
```
Comment (qingqing01) [inline paddle/operators/crop_op.cc]: LoDTensor合入后，InferShape的Output里要使用：`Output<framework::LoDTensor>`

File: paddle/operators/crop_op.cc [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("X"), "Input(X) should not be null");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar(framework::GradVarName("Out")),
+                            "Input(Out@GRAD) should not be null");
+    auto x_dims = ctx.Input<Tensor>("X")->dims();
+    auto *x_grad = ctx.Output<Tensor>(framework::GradVarName("X"));
```
Comment (qingqing01) [inline paddle/operators/crop_op.cc]: Output使用：`Output< framework ::LoDTensor>`

File: paddle/operators/crop_op.h [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        CropFunction<Place, T, 4>(context);
+        break;
+      case 5:
+        CropFunction<Place, T, 5>(context);
+        break;
+      case 6:
+        CropFunction<Place, T, 6>(context);
+        break;
+      default:
+        LOG(ERROR) << "Only ranks up to 6 supported.";
```
Comment (qingqing01) [inline paddle/operators/crop_op.h]: 报错不要使用GLOG，使用PADDLE_ENFORCE/PADDLE_THROW:

```cpp
PADDLE_THROW("Tensors with rank at most 6 are supported").
```

File: paddle/operators/crop_op.h [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+        CropGradFunction<Place, T, 4>(context);
+        break;
+      case 5:
+        CropGradFunction<Place, T, 5>(context);
+        break;
+      case 6:
+        CropGradFunction<Place, T, 6>(context);
+        break;
+      default:
+        LOG(ERROR) << "Only ranks up to 6 supported.";
```
Comment (qingqing01) [inline paddle/operators/crop_op.h]: 同上。

File: python/paddle/v2/framework/tests/test_crop_op.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    def setUp(self):
+        self.type = "crop"
+        self.inputs = {'X': np.random.random((16, 16)).astype("float32"), }
+        self.attrs = {}
+        self.attrs['offsets'] = [2, 3]
+        self.attrs['shape'] = [8, 8]
+        self.outputs = {'Out': self.inputs['X'][2:10, 3:11]}
+
+
+class TestCropGradOp(GradientChecker):
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_crop_op.py]: 使用新的单测框架。

File: python/paddle/v2/framework/tests/test_crop_op.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    def setUp(self):
+        self.op = Operator(
+            type="crop", X="X", Out="Out", offsets=[2, 3], shape=[8, 8])
+        self.inputs = {'X': np.random.random((16, 16)).astype("float32"), }
+
+    def test_normal(self):
+        self.check_grad(
+            self.op, self.inputs, set(["X"]), "Out", max_relative_error=0.5)
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_crop_op.py]: max_relative_error=0.5 太大

File: paddle/operators/crop_op.h (hunk 2 above)
Comment (wanghaoshuang) [reply paddle/operators/crop_op.h]: Fixed.

File: paddle/operators/crop_op.cc (hunk 4 above)
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cc]: Thx. Fixed.

File: paddle/operators/crop_op.cc (hunk 5 above)
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cc]: Fixed.

File: paddle/operators/crop_op.cc (hunk 6 above)
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cc]: Fixed.

File: paddle/operators/crop_op.h (hunk 8 above)
Comment (wanghaoshuang) [reply paddle/operators/crop_op.h]: Fixed.

File: python/paddle/v2/framework/tests/test_crop_op.py (hunk 10 above)
Comment (wanghaoshuang) [reply python/paddle/v2/framework/tests/test_crop_op.py]: Fixed.

File: paddle/operators/crop_op.h (hunk 9 above)
Comment (wanghaoshuang) [reply paddle/operators/crop_op.h]: Fixed.

File: python/paddle/v2/framework/tests/test_crop_op.py (hunk 11 above)
Comment (wanghaoshuang) [reply python/paddle/v2/framework/tests/test_crop_op.py]: Fixed.

File: paddle/operators/crop_op.cc (hunk 7 above)
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cc]: Fixed.

File: paddle/operators/crop_op.cc [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+  }
+  return result;
+}
+
+template <typename T>
+class CropCPUKernel : public framework::OpKernel {
+ public:
+  void Compute(const framework::ExecutionContext &context) const override {
+    auto *x = context.Input<LoDTensor>("X");
+    auto *out = context.Output<LoDTensor>("Out");
```
Comment (qingqing01) [inline paddle/operators/crop_op.cc]: 在非sequence operators里的OpKernel里实际可以不使用LoDTensor，用Tensor即可：

https://github.com/PaddlePaddle/Paddle/issues/4151

File: paddle/operators/crop_op.cc [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+    auto x_data = x->data<T>();
+    T *out_data = out->mutable_data<T>(context.GetPlace());
+    auto x_dims = x->dims();
+    auto out_dims = out->dims();
+    int64_t out_count = framework::product(out_dims);
+    std::vector<int64_t> x_shape = framework::vectorize(x_dims);
+    std::vector<int64_t> out_shape = framework::vectorize(out_dims);
+
+    auto offsets = context.op().Attr<std::vector<int>>("offsets");
```
Comment (qingqing01) [inline paddle/operators/crop_op.cc]: `context.op().Attr<> -> context.Attr<>`

File: paddle/operators/crop_op.cu [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+void CropCUDAFunctoin(const framework::ExecutionContext& context) {
+  PADDLE_ENFORCE(platform::is_gpu_place(context.GetPlace()),
+                 "It must use GPUPlace.");
+  auto* x = context.Input<LoDTensor>("X");
+  auto* out = context.Output<LoDTensor>("Out");
+  auto x_data = x->data<T>();
+  T* out_data = out->mutable_data<T>(paddle::platform::GPUPlace());
+  auto x_dims = x->dims();
+  auto out_dims = out->dims();
+  int64_t out_count = framework::product(out_dims);
```
Comment (qingqing01) [inline paddle/operators/crop_op.cu]: use `Tensor::numel()` instead of `framework::product`.

File: paddle/operators/crop_op.cu [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+    x_shape[i] = x_dims[i];
+    out_shape[i] = out_dims[i];
+  }
+  int64_t* x_shape_gpu;
+  int64_t* out_shape_gpu;
+  cudaMalloc((void**)&x_shape_gpu, sizeof(int64_t) * D);
+  cudaMemcpy(x_shape_gpu, x_shape, sizeof(int64_t) * D, cudaMemcpyHostToDevice);
+  cudaMalloc((void**)&out_shape_gpu, sizeof(int64_t) * D);
+  cudaMemcpy(out_shape_gpu, out_shape, sizeof(int64_t) * D,
+             cudaMemcpyHostToDevice);
```
Comment (qingqing01) [inline paddle/operators/crop_op.cu]: Do not use `cudaMalloc` and `cudaMemcpy`, use Tensor's function.

File: paddle/operators/crop_op.cc [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    auto x_dim = ctx.Input<LoDTensor>("X")->dims();
+    auto Y = ctx.Input<LoDTensor>("Y");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("X"),
+                            "Input(X) of CropOp should not be null.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.OutputVar("Out"),
+                            "Output(Out) of CropOp should not be null.");
```
[the remark this answers is not in the document]
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cc]: Fixed.

File: paddle/operators/crop_op.cc [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+    if (Y == nullptr) {
+      auto shape = Attr<std::vector<int>>("shape");
+      PADDLE_ENFORCE_EQ(
+          int64_t(shape.size()), x_dim.size(),
+          "Shape size should be equal to dimention size of input tensor.");
+      std::vector<int64_t> tensor_shape(shape.size());
+      for (size_t i = 0; i < shape.size(); ++i) {
+        tensor_shape[i] = (int64_t)shape[i];
```
[the remark this answers is not in the document]
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cc]: Fixed.

File: paddle/operators/crop_op.cc [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+               The size of shape list should be as same as 
+               dimension size of  input X.
+
+The input should be a k-D tensor(k > 0 and k < 7). As an example:
+
+Given:
+
+X = [[0, 1, 2, 0, 0]
+       [0, 3, 4, 0, 0]
+       [0, 0, 0, 0, 0]]
```
[the remark this answers is not in the document]
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cc]: Fixed.

File: paddle/operators/crop_op.cc [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+    AddAttr<std::vector<int>>("offsets",
+                              "A list<int> describing offsets to be cropped."
+                              "The size of offsets list should be as same as "
+                              "dimension size of  input X.");
+    AddAttr<std::vector<int>>("shape",
+                              "A list<int> describing the shape of output."
+                              "The size of shape list should be as same as "
+                              "dimension size of  input X.")
+        .SetDefault(std::vector<int>());
+  }
```
[the remark this answers is not in the document]
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cc]: Fixed.

File: paddle/operators/crop_op.cc (hunk 12 above)
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cc]: Fixed.

File: paddle/operators/crop_op.cc [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+class CropCPUKernel : public framework::OpKernel {
+ public:
+  void Compute(const framework::ExecutionContext &context) const override {
+    auto *x = context.Input<LoDTensor>("X");
+    auto *out = context.Output<LoDTensor>("Out");
+    auto x_data = x->data<T>();
+    T *out_data = out->mutable_data<T>(context.GetPlace());
+    auto x_dims = x->dims();
+    auto out_dims = out->dims();
+    int64_t out_count = framework::product(out_dims);
```
[the remark this answers is not in the document]
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cc]: Fixed.

File: paddle/operators/crop_op.cc (hunk 13 above)
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cc]: Fixed.

File: paddle/operators/crop_op.cu (hunk 14 above)
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cu]: Fixed.

File: paddle/operators/crop_op.cu (hunk 15 above)
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cu]: Fixed.

**Comment by reyoung** [comment]
@wanghaoshuang  I just push PR #4174, which will make this PR not to use `const_cast`. Please update.

Also, I think maybe `Transform` is better to implement `Crop`.

* https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/platform/transform.h

**Comment by reyoung** [comment]
I will give a try and submit a PR to you.

**Comment by reyoung** [comment]
Sorry, I misunderstand what `crop` means. It seems that it is not related to `Transform` at all.

**Comment by reyoung** [comment]
Please reference #4205. I think Crop and Concat can be abstracted by `StridedMemcpy`

File: paddle/operators/crop_op.cc [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+      std::vector<int64_t> tensor_shape(shape.size());
+      for (size_t i = 0; i < shape.size(); ++i) {
+        tensor_shape[i] = static_cast<int64_t>(shape[i]);
+      }
+      ctx.Output<LoDTensor>("Out")->Resize(framework::make_ddim(tensor_shape));
+    } else {
+      PADDLE_ENFORCE_EQ(framework::arity(x_dim), framework::arity(Y->dims()),
+                        "Tensor rank of both CropOp's "
+                        "inputs must be same.");
+      ctx.Output<LoDTensor>("Out")->Resize(Y->dims());
```
Comment (dzhwinter) [inline paddle/operators/crop_op.cc]: use `ctx.Output<LodTensor>` twice, it's better to reuse.

File: paddle/operators/crop_op.cu [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+
+  Tensor crop_rules_gpu;
+  crop_rules_gpu.CopyFrom<int>(crop_rules, paddle::platform::GPUPlace());
+
+  int n = out_dims[0];
+  int d = out_dims[1];
+  int block = 512;
+  int grid = (n * d + block - 1) / block;
+
+  CropKernel<
```
Comment (dzhwinter) [inline paddle/operators/crop_op.cu]: is it re-fold by the pre-commit? The format is strange, maybe it's better to add a new line character in next line.

File: python/paddle/v2/framework/tests/test_crop_op.py [hunk 23]
```diff
@@ -0,0 +1,91 @@
+import unittest
```
Comment (dzhwinter) [inline python/paddle/v2/framework/tests/test_crop_op.py]: the copyrights.

File: paddle/operators/crop_op.cc (hunk 21 above)
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cc]: Fixed.

File: paddle/operators/crop_op.cc [hunk 24]
```diff
@@ -0,0 +1,186 @@
...
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("X"),
+                            "Input(X) of CropOp should not be null.");
+    PADDLE_ENFORCE_NOT_NULL(ctx.OutputVar("Out"),
+                            "Output(Out) of CropOp should not be null.");
+    auto x_dim = ctx.Input<LoDTensor>("X")->dims();
+    auto Y = ctx.Input<LoDTensor>("Y");
```
[the remark this answers is not in the document]
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cc]: Fixed.

File: python/paddle/v2/framework/tests/test_crop_op.py (hunk 23 above)
Comment (wanghaoshuang) [reply python/paddle/v2/framework/tests/test_crop_op.py]: 好像所有单测都没copyrights。确定需要加上么？

File: paddle/operators/crop_op.cu (hunk 22 above)
Comment (wanghaoshuang) [reply paddle/operators/crop_op.cu]: Fixed.

Comment (wanghaoshuang) [reply paddle/operators/crop_op.cu]: Fixed.

**Pull request merged**